include/swift/SIL/Mangle.h

```diff
@@ -104,6 +104,8 @@ template <typename SubType>
 class SpecializationMangler : public SpecializationManglerBase {
   SubType *asImpl() { return static_cast<SubType *>(this); }
 public:
+  CanSILFunctionType CanSILFnTy;
```

Comment:
I'm not sure it is a hack. With substitutions containing archetype, we want to mangle the actual SIL function type of the specialized function. This type cannot be easily derived by the `SpecializationMangler` from the original generic function and a set of substitutions. This is why it takes a SIL function type of the new function now.